proposals/NNNN-shader-execution-reordering.md

```diff
@@ -0,0 +1,1388 @@
+- `HitObject::Invoke`: transitions to and from `closeshit` and `miss` shaders.
+Constitutes a reorder point even in cases where no shader is invoked.
+- `ReorderThread`: the `ReorderThread` call site.
+
+`ReorderThread` stands out as it explicitly separates reordering from a
+transition between shader stages, thus, it allows applications to (carefully)
+choose the most effective reorder locations given a specific workload. The
+combination of `HitObject` and coherence hints provides additional control
+over the reordering itself. These characteristics make `ReorderThread` a
+versatile tool for improving performance in a variety of workloads that suffer
+from divergent execution or data access.
+
+Similar to `TraceRay` and `CallShader`, the behavior of the actual reorder
+operation at any reorder point is implementation specific. An implementation
+may, for instance, not reorder at all, reorder within some local domain
+(threads on a local processor), reorder only within a certain time window,
+reorder globally across the entire GPU and the entire dispatch, or any
+variation thereof. The general target is for implementations to make a _best
+effort_ to extract as much coherence as possible, while keeping the overhead
+of reordering low enough to be practical for use in typical raytracing
+scenarios.
+
+---
+
+### Divergence behavior
+
+An implementation may only reorder threads whose control flow arrives at a
+reorder point. Threads that do not arrive at a reorder point are guaranteed to
+not migrate, even if neighboring threads in the same wave do migrate.
+
+From the physical wave's point of view, an implementation may remove, replace,
+or do nothing with threads that arrive at a reorder point. It is also legal
+for an implementation to replace terminated threads in a wave if any of its
+threads arrive at a reorder point. This may indirectly affect threads that
+conditionally did not invoke a reorder point, as illustrated in the following
+code snippet:
+
+```C++
+int MyFunc(int coherenceCoord)
+{ 
+    int A = WaveActiveBallot(true); 
+    if (WaveIsFirstLane()) 
+        ReorderThread(coherenceCoord, 32); 
+    int B = WaveActiveBallot(true); 
+    return A - B; 
+}
+```
+
+In this example, a number of different things could happen:
+- If the implementation does not honor `ReorderThread` at all, the function
+will most likely return zero, as the set of threads before and after the
+conditional reorder would be the same. 
+- If the implementation reorders threads invoking `ReorderThread` but does not
+replace them, B will likely be less than A for threads not invoking
+`ReorderThread`, while the reordered threads will likely resume execution with
+a newly formed full wave, thereby obtaining `A <= B`.
+- If the implementation replaces threads in a wave, the threads not
+participating in the reorder may possibly be joined by more threads than were
+removed from the wave, again observing `A <= B`.
```

Comment:
What determines these different possible outcomes? Is it implementation or situation dependent or both?

Reply:
It could both be up to the implementation and situation.